File: 57-insert-interval/insert-interval.js
/**
 * @param {number[][]} intervals
 * @param {number[]} newInterval
 * @return {number[][]}
 */
var insert = function(intervals, newInterval) {
    // loop through intervals
    // newInterval end < interval[0]
        // add in new interval and return
    // newINterval end > interval[1]
        // add inteval to new arr and continue
    // update newInterval with new minmax
    const output = []

    for(let i = 0; i < intervals.length; i++) {
        if(newInterval[1] < intervals[i][0]) {
            output.push(newInterval)
            output.push(...intervals.slice(i))
            return output
        } else if (newInterval[0] > intervals[i][1]) {
            output.push(intervals[i])
        } else {
            newInterval = [
                Math.min(newInterval[0], intervals[i][0]),
                Math.max(newInterval[1], intervals[i][1]),
            ]
        }
    }

    output.push(newInterval)
    return output
};

File: 57-insert-interval/insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        # non overlapping
        # loop through array
        # if new interval end is less than interval[i]start
        # - add in interval
        # - add rest of intervals and return
        # if new interval start > interval[i] end
        # - continue
        # last one is overlapping so merge

        # time complexity 0(N)

        arr = []

        for i in range(len(intervals)):
            if newInterval[1] < intervals[i][0]:
                arr.append(newInterval)
                arr.extend(intervals[i:])
                return arr
            elif newInterval[0] > intervals[i][1]:
                arr.append(intervals[i])
            else:
                newInterval = [
                    min(newInterval[0], intervals[i][0]),
                    max(newInterval[1], intervals[i][1])
                ]
        
        arr.append(newInterval)
        return arr